### ISSUE TITLE
Invalid warning on `redundant_pattern_matching`

### ISSUE THREAD

**Issue by nurmohammed840** [rust-lang/rust-clippy#11174]
### Summary

This  example should not yield any warning. 

I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b6aca8a9ddbf93f6ff01a02a1ec8467e):

```rust
fn _test<T>(boolean: bool, maybe_some: Option<T>) -> bool {
    matches!(maybe_some, None if !boolean)
}
```

I saw this happen:

```
warning: redundant pattern matching, consider using `is_none()`
 --> src/lib.rs:2:5
  |
2 |     matches!(maybe_some, None if !boolean)
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `maybe_some.is_none()`
  |
```

### Lint Name

redundant_pattern_matching

### Reproducer

I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b6aca8a9ddbf93f6ff01a02a1ec8467e):

```rust
fn _test<T>(boolean: bool, maybe_some: Option<T>) -> bool {
    matches!(maybe_some, None if !boolean)
}
```

I saw this happen:

```
warning: redundant pattern matching, consider using `is_none()`
 --> src/lib.rs:2:5
  |
2 |     matches!(maybe_some, None if !boolean)
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `maybe_some.is_none()`
  |
```

### Version

```text
rustc 1.73.0-nightly (33a2c2487 2023-07-12)
```
